ddtrace/internal/wrapping/context.py

```diff
@@ -0,0 +1,686 @@
+        return value
+
+    @classmethod
+    def is_wrapped(cls, _f: FunctionType) -> bool:
+        raise NotImplementedError
+
+    @classmethod
+    def extract(cls, _f: FunctionType) -> "BaseWrappingContext":
+        raise NotImplementedError
+
+    def wrap(self) -> None:
+        raise NotImplementedError
+
+    def unwrap(self) -> None:
+        raise NotImplementedError
+
+
+# This is the public interface exported by this module
+class WrappingContext(BaseWrappingContext):
+    @override
+    def __enter__(self, _frame: FrameType) -> "WrappingContext":
+        raise NotImplementedError
+
+    @classmethod
+    def is_wrapped(cls, f: FunctionType) -> bool:
+        try:
+            return bool(cls.extract(f))
+        except ValueError:
+            return False
+
+    @classmethod
+    def extract(cls, f: FunctionType) -> "WrappingContext":
+        if _UniversalWrappingContext.is_wrapped(f):
+            try:
+                return _UniversalWrappingContext.extract(f).registered(cls)
+            except KeyError:
+                pass
+        msg = f"Function is not wrapped with {cls}"
+        raise ValueError(msg)
+
+    def wrap(self) -> None:
+        t.cast(_UniversalWrappingContext, _UniversalWrappingContext.wrapped(self.__wrapped__)).register(self)
+
+    def unwrap(self) -> None:
+        f = self.__wrapped__
+
+        if _UniversalWrappingContext.is_wrapped(f):
+            _UniversalWrappingContext.extract(f).unregister(self)
+
+
+class ContextWrappedFunction(Protocol):
+    """A wrapped function."""
+
+    __dd_context_wrapped__ = None  # type: t.Optional[_UniversalWrappingContext]
+
+    def __call__(self, *args, **kwargs):
+        pass
+
+
+# This class provides an interface between single bytecode wrapping and multiple
+# logical context wrapping
+class _UniversalWrappingContext(BaseWrappingContext):
+    def __init__(self, f: FunctionType) -> None:
+        super().__init__(f)
+
+        self._contexts: t.Dict[t.Type[WrappingContext], WrappingContext] = {}
+
+    def register(self, context: WrappingContext) -> None:
+        _type = type(context)
+        if _type in self._contexts:
+            raise ValueError("Context already registered")
+
+        self._contexts[_type] = context
+
+    def unregister(self, context: WrappingContext) -> None:
+        _type = type(context)
+        if _type not in self._contexts:
+            raise ValueError("Context not registered")
+
+        del self._contexts[_type]
+
+        if not self._contexts:
+            self.unwrap()
+
+    def is_registered(self, context: WrappingContext) -> bool:
+        return type(context) in self._contexts
+
+    def registered(self, context_type: t.Type[WrappingContext]) -> WrappingContext:
+        return self._contexts[context_type]
+
+    def __enter__(self) -> "_UniversalWrappingContext":
+        frame = sys._getframe(1)
+        for context in sorted(self._contexts.values(), key=lambda c: c.__priority__):
```

Comment:
should we just store these in a sorted dict and we can resort on insert to avoid this sort here?

also, instead of a dict, it might just be easier to store as a list or tuple, and on insert you iterate the whole list to see if the type is already present.

_might_ be lower overhead than using a dict here? I mostly wonder if we can get the per-call overhead down as low as possible, especially since the registering should be called very infrequently, at least compared to `__enter__`/`__exit__`/`__return__`.

Reply:
My thinking was that we shouldn't really have that many things registered here, so sorting and traversing should be quick enough. If we are really concerned about performance then the better data structure would probably be a heap, but we would have to back that with a dict anyway, to keep track of what's registered